### PR TITLE
feat(mito): avoid buffering all batches for the same primary key

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR implements the merge reader in a more streaming way. The merge reader checks whether the buffered rows of the batch are more than `batch_size` and outputs the batch if it exceeds `batch_size`. This may slow down the reader but the query engine may benefit from a larger batch.

It adds more tests to check the correctness of this behavior.

https://github.com/GreptimeTeam/greptimedb/issues/2660 describes the merge algorithm.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
https://github.com/GreptimeTeam/greptimedb/issues/2660

### CODE REVIEW

**Review by fengjiachun — APPROVED** [review]
DLJB

File: src/mito2/src/read.rs [hunk 1]
```diff
@@ -647,6 +647,17 @@ impl Source {
             Source::Iter(iter) => iter.next().transpose(),
         }
     }
+
+    /// Returns next non-empty [Batch].
+    pub(crate) async fn next_non_empty_batch(&mut self) -> Result<Option<Batch>> {
```
Comment (waynexia) [inline src/mito2/src/read.rs:652]: `next_batch` should be enough. The return type is `Option` which includes empty and non-empty cases

Comment (evenyag) [reply src/mito2/src/read.rs:652]: In most cases, it should be enough, but the `Batch` itself may be `empty` (no rows in the batch).

Comment (evenyag) [reply src/mito2/src/read.rs]: For simplicity, I remove this method.

**Review by waynexia — APPROVED** [review]
kudu master 🦌

**Pull request merged**